### PR TITLE
better support for special characters in DATABASE_URL

### PR DESCRIPTION
**Pull request by robjuz**
## Description
This is a fix for the previous implemetation with handling for the '+' charakter

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [x] I verified that my code applies to the guidelines (`composer code-check`)
- [ ] I updated the documentation (see [here](https://github.com/kimai/www.kimai.org/tree/master/_documentation))
- [x] I agree that this code is used in Kimai (see [license](https://github.com/kimai/kimai/blob/main/LICENSE))

### CODE REVIEW

**Comment by kevinpapst** [comment]
I don't get why you use `urldecode`. Can you explain?

**Comment by robjuz** [comment]
@kevinpapst Hi. This implementation handles the `+` sign problem. It looks a little bit dirty, but I couldn't find a better solution.

 I tested it with following configuration

 `database setup`
 ```.env
 MYSQL_DATABASE: kimai+database
 MYSQL_USER: kimai+user
 MYSQL_PASSWORD: kimai?password
 ```
 `.env`
 ```
 DATABASE_URL=mysql://kimai+user:kimai%3Fpassword@localhost:3306/kimai%2Bdatabase?charset=utf8&serverVersion=5.7
 ```

 You can see that the `+` is users as plain test in the `user` and urlencoded in the `database`
 
 The interesting part of the startup script
 
 ```bash
 Wait for MySQL DB connection ...
 + DB_BASE=kimai%2Bdatabase
 + re='^[0-9]+$'
 + [[ 3306 =~ ^[0-9]+$ ]]
 + echo 'Wait for MySQL DB connection ...'
 + php /dbtest.php localhost kimai%2Bdatabase 3306 kimai+user kimai%3Fpassword
 + echo 'Connection established'
 + handleStartup
 + /opt/kimai/bin/console -n kimai:install
 Testing DB:** new \PDO(mysql:host=localhost;dbname=kimai+database;port=3306, kimai+user, kimai?password, [ \PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION ]);*Connection established
 ```

**Comment by kevinpapst** [comment]
Aha, so the user added a correct encoded URL and Kimai works, but only this test script failed?

**Comment by robjuz** [comment]
According to Symfony documentation: 
>If the username, password, host or database name contain any character considered special in a URI (such as : / ? # [ ] @ ! $ & ' ( ) * + , ; =), you must encode them. See RFC 3986 for the full list of reserved characters or use the urlencode function to encode them. In this case you need to remove the resolve: prefix in config/packages/doctrine.yaml to avoid errors: url: '%env(DATABASE_URL)%'

So I `urlencode` the `DATABASE_URL` for symfony and for the `dbtest.php` I `urldecode` it.

**Comment by robjuz** [comment]
> Aha, so the user added a correct encoded URL and Kimai works, but only this test script failed?

Exactly. And the `+` was handled incorrectly

**Comment by kevinpapst** [comment]
Hm... this will still cause issues if someone uses a password like `mG0d1 3aTZ+s1`.

The idea is that you have to `urlencode()` credentials with special character:

```bash
php -r "echo urlencode('mG0d1 3aTZ+s1');"

mG0d1+3aTZ%2Bs1
```

If you add this to your `DATABASE_URL` the test script will now break the password.
Maybe I reverted yesterday to quickly and it was actually [the users fault](https://github.com/kimai/kimai/issues/4997) to not urlencode their password...

**Comment by robjuz** [comment]
hmm... It looks like we have 2 options.

1. Drop this feature
2. Go back to my first implementation (that looks correct to my) and make it a braking change.

**Comment by kevinpapst** [comment]
I need to do some testing. But I think we can safely assume that the credentials are correctly urlencoded when they go into the dbtest script.

> Go back to my first implementation

Yes, I was tired an d a bit sick the last days, should not have done a rollback so quickly...

Let me do some local tests and then I will report back.

**Comment by robjuz** [comment]
Sure. No stress. The idea for the PR came from an issue created in my helm chart repo long time ago and I decided now to take a look on it. It can definitely wait a little longer ;)

**Comment by kevinpapst** [comment]
Let's go with your initial PR. This workaround looks wrong to me. 

Sorry, my fault for reverting! 

On a side note, do you have experience or any feedback on these two docker PRs: 
- https://github.com/kimai/kimai/pull/4995
- https://github.com/kimai/kimai/pull/4996

I don't have the confidence yet to merge container changes ...

File: .docker/dbtest.php
Comment (kevinpapst) [inline .docker/dbtest.php]:
```suggestion
$DB_HOST = urldecode($argv[1]);
$DB_BASE = urldecode($argv[2]);
```

File: .docker/dbtest.php
Comment (kevinpapst) [inline .docker/dbtest.php]:
```suggestion
$DB_USER = urldecode($argv[4]);
$DB_PASS = urldecode($argv[5]);
```

**Review by kevinpapst — APPROVED** [review]

**Comment by robjuz** [comment]
@kevinpapst I will take a look on the docker part of this project. Maybe I can provide some improvements.

**Comment by kevinpapst** [comment]
Thank you 👍 

Toby Batch, the original docker maintainer, doesn't have the time to always support and I am not very experienced with Docker. So a second pair of eyes is very welcome!

**Pull request merged**
